### DISCUSSION TITLE
Build failed for Swift 5.3 using Xcode 12.4, have no idea why

### DISCUSSION THREAD

**Discussion by Janneman84** [SwiftPackageIndex/SwiftPackageIndex-Server#1781] [Q&A] [answered]
My package fails to build on the Swift 5.3 test and I'm looking for help. The report is here:
https://swiftpackageindex.com/builds/781136EE-2076-4691-BCFD-119C733E2F6F

I even downloaded Xcode 12.4 to see for myself but when I try it seems to build fine.

I get this generic error:
`error: failed to produce diagnostic for expression; please file a bug report`

About this line of code:
```
DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak viewController, weak subview] in
```
Any Swift experts that can tell me why this would fail on Swift 5.3? Thanks!

**Comment by finestructure** [comment]
Sorry to hear you're having trouble with the build system!

I'm a bit confused - you say it seems to build fine but you're getting a generic error?

Does the actual build command:

```
env DEVELOPER_DIR="/Applications/Xcode-12.4.0.app" xcrun xcodebuild -IDEClonedSourcePackagesDirPathOverride="$PWD/.dependencies" -derivedDataPath "$PWD/.derivedData" build -scheme "LeakedViewControllerDetector" -destination "generic/platform=ios"
```

work for you?

Unfortunately, I can't easily test builds with 12.4 on my machine but we'll try and reproduce it ourselves as well.

**Reply by Janneman84** [reply]
The error is from your build system, if you open the link you can see the details.

I downloaded and started Xcode 12.4, created a new project and added my package. It builds and doesn't complain whatsoever.

I don't use the build command because it won't help me solve the problem. I mean either it won't give an error, which isn't helpful, or or it will show the generic error which isn't helpful either.

I was hoping that the Xcode editor would point out a more clear error but it shows none. I also tried the code on [swiftfiddle.com](https://swiftfiddle.com) but you can't use DispatchQueue stuff there.

**Reply by finestructure** [reply]
> I don't use the build command because it won't help me solve the problem. I mean either it won't give an error, which isn't helpful, or or it will show the generic error which isn't helpful either.

Please give it a try though, since you've got Xcode 12.4 already - it'll tell us if it's a problem with our build setup or something we misinterpret when trying to build your project. It's the easiest way for us to help you resolve this problem :)

**Reply by Janneman84** [reply]
I'm trying but whatever I try it just says
`xcrun: error: missing DEVELOPER_DIR path: /Downloads/Xcode.app`

I've tried with ~ and other things but no luck.

**Reply by finestructure** [reply]
You need to substitute in the path where you installed Xcode 12.4. That's probably `/Application/Xcode-12.4`. You can run

```
ls /Application/Xcode*
```

to verify what it's called exactly. It should list all the Xcodes in your Application folder.

**Reply by finestructure** [reply]
Note that if the path has a space in it, it's important you keep the quotes in the command, like so:

```
env DEVELOPER_DIR="/Applications/Xcode 12.4.app" ...
```

**Reply by Janneman84** [reply]
Yes I finally got it working after apparently some silly mistake.

It does trigger the error here too. After some experimenting it seems to have an issue with adding a weak declaration of a callback var:

```swift
view.iterateTopSubViews(viewController: viewController) { subview in
    DispatchQueue.main.asyncAfter(deadline: .now() + delay) { [weak viewController, weak subview] in
```

As you can see `subview` is a callback var that is passed on as weak in the other callback of the next line, which triggers the error. To fix I simply added this between the two lines:
```swift
let subview = subview
```

Everything seems to function fine with this somehow.

I also had to explicitly type `UIColor.systemPink` instead of the shorter `.systemPink`.

Thanks for sticking around to help me!

**Reply by finestructure** [reply]
No worries, glad you got it working!

**Comment by Janneman84** [comment]
After updating to Sonoma I am unable to run Xcode 12.4's CLI anymore, is this a known problem?

It's unable to find CoreChart or something.

`Library not loaded: /System/Library/PrivateFrameworks/CoreChart.framework/Versions/A/CoreChart`

Is it Sonoma or me?

**Reply by finestructure** [reply, marked as answer]
Sonoma doesn't support Xcode versions earlier than 15. You can get Xcode 14 to work with a little trick but it's not officially supported, not sure about versions even earlier than that. The only reliable way I know to run Xcode < 15 on Sonoma is to virtualise an older macOS version.

**Reply by Janneman84** [reply]
I realized SPI upped the lowest tested Swift version to 5.6. So I tried Xcode 13.3 instead, which works fine thanks!
